pd/src/consensus/worker.rs

```diff
@@ -247,164 +241,235 @@ impl Worker {
             .height
             .try_into()
             .expect("height should be nonnegative");
-        let epoch = pending_block.set_height(height);
+        let epoch = pending_block.set_height(
+            height,
+            self.state
+                .private_reader()
+                .chain_params_rx()
+                .borrow()
+                .epoch_duration,
+        );
 
         tracing::debug!(?height, ?epoch, end_height = ?epoch.end_height());
 
+        // Find out which validators were slashed in this block
+        let slashed_validators = pending_block
+            .validator_state_changes
+            .iter()
+            .filter(|p| matches!(p.1, ValidatorState::Slashed))
+            .map(|p| p.0) // get the validator ID
+            .collect::<Vec<_>>();
+
+        // Immediately revert notes and nullifiers immediately from slashed validators in this block
+        let (mut slashed_notes, mut slashed_nullifiers) = (
+            reader.quarantined_notes(None, Some(slashed_validators.iter())),
+            reader.quarantined_nullifiers(None, Some(slashed_validators.iter())),
+        );
+        while let Some(result) = slashed_notes.next().await {
+            pending_block.reverting_notes.insert(result?.1); // insert the commitment
+        }
+        while let Some(result) = slashed_nullifiers.next().await {
+            pending_block.reverting_nullifiers.insert(result?.1); // insert the nullifier
+        }
+        drop(slashed_notes);
+        drop(slashed_nullifiers);
+
+        // If we are at the end of an epoch, process changes for it
         if epoch.end_height().value() == height {
-            // We've finished processing the last block of `epoch`, so we've
-            // crossed the epoch boundary, and (prev | current | next) are:
-            let prev_epoch = epoch;
-            let current_epoch = prev_epoch.next();
-            let next_epoch = current_epoch.next();
-
-            tracing::info!(
-                ?height,
-                ?prev_epoch,
-                ?current_epoch,
-                ?next_epoch,
-                "crossed epoch boundary, processing rate updates"
-            );
-            metrics::increment_counter!("epoch");
+            self.end_epoch().await?;
```

Comment:
All the end-of-epoch code is factored out into this one function now :)

Reply:
Going forward, we'll need to change the signature of `end_epoch()` at some point so that we can propagate validator info into the `EndBlock` response.